src/main/java/org/jamesframework/core/search/NeighbourhoodSearch.java

```diff
@@ -367,6 +371,106 @@ protected boolean isImprovement(Move<? super SolutionType> move){
     }
     
     /**
+     * <p>
+     * Get the best valid move among a collection of possible moves. The best valid move is the FIRST one yielding the positive delta 
+     * or the one that is non-tabu and yield the largest delta(see {@link #computeDelta(Evaluation, Evaluation)}) when being applied to the current solution.
+     * </p>
+     * <p>
+     * If <code>requireImprovement</code> is set to <code>true</code>, only moves that improve the current solution
+     * are considered, i.e. moves that yield a positive delta (unless the current solution is invalid, then all
+     * valid moves are improvements). Any number of additional filters can be specified so that moves are only
+     * considered if they pass through all filters. Each filter is a predicate that should return <code>true</code>
+     * if a given move is to be considered. If any filter returns <code>false</code> for a specific move, this
+     * move is discarded.
+     * </p>
+     * <p>
+     * Returns <code>null</code> if no move is found that satisfies all conditions.
+     * </p>
+     * <p>
+     * Note that all computed evaluations and validations are cached.
+     * Before returning the selected best move, if any, its evaluation and validity are cached
+     * again to maximize the probability that these values will remain available in the cache.
+     * </p>
+     * 
+     * @param moves collection of possible moves
+     * @param requireImprovement if set to <code>true</code>, only improving moves are considered
+     * @param acceptFirstImprovement if set to <code>true</code>, return the first improvement move
+     * @param filters additional move filters
+     * @return first improving move or the best valid move, may be <code>null</code>
+     */
+    @SafeVarargs
+    protected final Move<? super SolutionType> getBestMove(Collection<? extends Move<? super SolutionType>> moves,
+                                                           boolean requireImprovement,
+                                                           boolean acceptFirstImprovement,
+                                                           Predicate<? super Move<? super SolutionType>>... filters){
+        // track first improvement move AND best valid move + corresponding evaluation, validation and delta
+        Move<? super SolutionType> bestMove = null, firstImprovementMove = null;
+        double bestMoveDelta = -Double.MAX_VALUE, curMoveDelta, firstImprovementMoveDelta = -Double.MAX_VALUE;
+        Evaluation curMoveEvaluation, bestMoveEvaluation = null, firstImprovementMoveEvaluation = null;
```

Comment:
No need to store first and best move (and their metadata separately). You can have one series of variables for the chosen move, say `chosenMove`, `chosenMoveDelta`, etc. which will either be the first or best one depending on the parameter `acceptFirstImprovement`.